server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java

```diff
@@ -152,21 +149,29 @@ protected boolean hasOrganizationPermissionImpl(String organizationUuid, String
 
   @Override
   protected boolean hasProjectUuidPermission(String permission, String projectUuid) {
-    if (!projectPermissionsCheckedByUuid.contains(permission)) {
-      try (DbSession dbSession = dbClient.openSession(false)) {
-        // FIXME do not load all the authorized projects. It can be huge.
-        Collection<String> projectUuids = dbClient.authorizationDao().selectAuthorizedRootProjectsUuids(dbSession, getUserId(), permission);
-        addProjectPermission(permission, projectUuids);
-      }
+    if (permissionsByProjectUuid == null) {
+      permissionsByProjectUuid = HashMultimap.create();
     }
-    return projectUuidByPermission.get(permission).contains(projectUuid);
+    Set<String> permissions;
+    if (permissionsByProjectUuid.containsKey(projectUuid)) {
+      permissions = permissionsByProjectUuid.get(projectUuid);
```

Comment:
why not `return permissionsByProjectUuid.get(projectUuid).contains(permission)` and drop the `else` and the undefined variable?